rllib/agents/dqn/simple_q.py

```diff
@@ -215,3 +250,71 @@ def execution_plan(workers, config, **kwargs):
         )
 
         return StandardMetricsReporting(train_op, workers, config)
+
+    @ExperimentalAPI
+    def training_iteration(self) -> ResultDict:
+        """Simple Q training iteration function.
+
+        This replicates Simple Q's execution_plan behaviour, e.g.:
+        - For every time we (1) sample (MultiAgentBatch) from workers...
+            - (2) Concatenate freshly collected samples.
+            - (3) Store new samples in replay buffer.
+            - (4) Sample training batch (MultiAgentBatch) from replay buffer.
+            - (5) Learn on training batch.
+            - (6) Update target network every target_network_update_freq steps.
+        - (7) Return all collected metrics for the iteration.
+
+        Returns:
+            The results dict from executing the training iteration.
+        """
+        batch_size = self.config["train_batch_size"]
+        local_worker = self.workers.local_worker()
+
+        train_results = {}
+
+        # (1) Sample (MultiAgentBatch) from workers
+        new_sample_batches = synchronous_parallel_sample(self.workers)
+
+        # Update counters
+        self._counters[NUM_ENV_STEPS_SAMPLED] += sum(len(s) for s in new_sample_batches)
+        self._counters[NUM_AGENT_STEPS_SAMPLED] += sum(
+            len(s) if isinstance(s, SampleBatch) else s.agent_steps()
+            for s in new_sample_batches
+        )
+
+        # (2) Concatenate freshly collected samples
+        concatenated_samples = SampleBatch.concat_samples(new_sample_batches)
+        # (3) Store new samples in replay buffer
+        self.local_replay_buffer.add(concatenated_samples)
+
+        # This is done to fit the old execution plan behaviour
```

Comment:
1) Just curious: I see that this could be desirable (to mimic the old behavior). However, did we try not doing this loop and simply update only once per parallel rollout?

2) Above, do we really need to concat all the samples? Why not add each individual batch to the buffer in a loop (batches will be split by timesteps anyways, which means we would save the (unnecessary) concat step).

Reply:
1. Simply training once per rollout throws the algorithm off and learning tests fail. I can do this, but I would have to adjust parameters and the algorithm would just not be the same.
This is still related to your [earlier request](https://github.com/ray-project/ray/pull/22842#discussion_r832440472). My issue with this is that in the original execution plan we do one training steps per collected fragment. Or with the default parameters: 4 training steps per parallel rollout. 

For every time we sample a fragment from a rollout worker, we train once - in the execution plan and now in the training iteration fn. This is because of the "round robin Concurrently operator" in the execution plan.

Maybe there is a misunderstanding here: I got rid of the dependency on the batch size only and left the loop there because I thought I should stick to the old behaviour and the old parameters, just like I would in other algorithms.